Generic Host PCIe Controller Support

superna9999 opened the pull request:
This PR is the first step into PCIe Host Controller support on non-x86 platforms (aka Generic or platform-agnostic)

This adds :
- edtlib/gen_defines: changes to handle the ranges DT attribute
- pci-host-ecam-generic bindings
- PCIe Controller API
- Generic PCIe Controller layer implementing the current PCIe API
- Generic PCIe Controller in ECAM mode driver

The Generic PCIe Controller layer provides:
- Configuration space read/write
- single bus endpoint enumerations
- Endpoint I/O, MEM & MEM64 BARs allocation
- Endpoint I/O, MEM & MEM64 BARs get & translation for drivers

The Generic PCIe Controller in ECAM mode driver provides:
- Raw DT RANGES properties into usable PCIe regions
- Configuration space read/write into ECAM config space
- PCIe regions allocation & translation

The limitations are:
- No support for PCIe prefetchable regions
- No support for PCIe bus configuration (only bus0 is supported)
- No support for multiple controllers (no domain-id in BDF)

Support has been designed to initially support Root Complexes with Root Complex Integrated Endpoint, which was designed for Embedded Systems with internal-only PCIe Endpoints on bus 0.

Testing has been done with QEMU with endpoints only on bus0, this can be achieved with:
```
-device <device>,bus=pcie.0
```

MSI/MSI-X support depends on #37506 completion

DT RANGES tooling and macros are pushed to #38106 and this serie is rebased on it, will need to be merged before this one.

Extended PCI(e) capability offset & defines changes were moved to #38256

superna9999 committed:
dts: arm64: qemu-virt: switch to 64bit addressing in DT

The qemu arm64 virt machine PCIe controller config space needs a
64bit adressing, so switch the entire qemu-virt-a53 and
qemu_cortex_a53 board dts to address-cells & size-cells to 2
to accommodate for 64bit addresses.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
pcie: add initial controller support

This adds :
- Generic PCIe Controller layer implementing the current PCIe API
- Generic PCIe Controller in ECAM mode driver

The Generic PCIe Controller layer provides:
- Configuration space read/write
- single bus endpoint enumerations
- Endpoint I/O, MEM & MEM64 BARs allocation
- Endpoint I/O, MEM & MEM64 BARs get & translation for drivers

The Generic PCIe Controller in ECAM mode driver provides:
- Raw DT RANGES properties into usable PCIe regions
- Configuration space read/write into ECAM config space
- PCIe regions allocation & translation

The limitations are:
- No support for PCIe prefetchable regions
- No support for PCIe bus configuration (only bus0 is supported)
- No support for multiple controllers (no domain-id in BDF)

Support has been designed to initially support Root Complexes with
Root Complex Integrated Endpoint, which was designed for Embedded
Systems with internal-only PCIe Endpoints on bus 0.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
bindings: pcie: add pci-host-ecam-generic bindings

Add bindings for a generic PCIe Controller in ECAM mode.

ECAM stands for PCI Express Enhanced Configuration Access Mechanism
where the PCIe config space is flat memory-mapped.

An optional msi-parent phandle is added to link to an optional
MSI/MSI-X Interrupt Message Translation HW component.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
pcie: add PCIe Controller API

This adds a generic PCIe Controller driver API providing the
necessary callbacks & config structure to handle the PCIe
config space and BAR regions handling.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

galak commented:
@superna9999 do you mind pull the devicetree tooling stuff into its own PR for review and acceptance.

galak commented on include/drivers/pcie/controller.h:
```diff
@@ -0,0 +1,340 @@
+/*
+ * Get BAR regions translation ranges count from DT
+ */
+#define DT_RANGES_COUNT(node_id) \
```
These should be in an include/devicetree/ header not here.

galak reviewed (dismissed):
* Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
* Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

superna9999 commented:
> * Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
> * Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

Will do in 2w after my vacations

tbursztyka commented on drivers/pcie/host/core.c:
```diff
@@ -0,0 +1,215 @@
+/*
```
same as for the file name, controller.c would be better no?

superna9999 replied:
Yes

superna9999 commented on drivers/pcie/host/Kconfig:
```diff
@@ -10,6 +10,23 @@ menuconfig PCIE
 
 if PCIE
 
+config PCIE_CORE
```
yes, I wasn't sure about naming, will rename to that

finikorg commented on include/drivers/pcie/msi.h:
```diff
@@ -111,8 +112,8 @@ extern bool pcie_msi_enable(pcie_bdf_t bdf,
-#define PCIE_MSI_CAP_ID		0x05U
-#define PCIE_MSIX_CAP_ID	0x11U
+#define PCIE_MSI_CAP_ID		PCI_CAP_ID_MSI
+#define PCIE_MSIX_CAP_ID	PCI_CAP_ID_MSIX
```
Can we just use `PCI_CAP_ID_MSI` instead of `PCIE_MSI_CAP_ID`.

superna9999 replied:
Obviously yes, I wanted to minimize changes but yes those defines should be dropped

tbursztyka commented on drivers/pcie/host/msi.c:
```diff
@@ -6,6 +6,7 @@
 
 #include <kernel.h>
 #include <drivers/pcie/msi.h>
+#include <drivers/pcie/cap.h>
```
Could you put these 3 last commits into their own PR? They are not specifically related to PCIe host controller support, and they could get merged faster as well.

mbolivar-nordic commented on include/drivers/pcie/controller.h:
```diff
@@ -0,0 +1,295 @@
+#define ZEPHYR_INCLUDE_DRIVERS_PCIE_CONTROLLERS_H_
+
+#include <zephyr/types.h>
+#include <devicetree/ranges.h>
```
Just include devicetree.h. All `devicetree/foo.h` headers are private.

superna9999 replied:
Yes, I need to rebase on #38106

mbolivar-nordic commented on drivers/pcie/host/controller.c:
```diff
@@ -0,0 +1,215 @@
+{
+	const struct device *dev;
+
+	dev = DEVICE_DT_GET(DT_ALIAS(pcie0));
```
This looks like it should be a chosen node. Aliases in Zephyr are generally reserved for samples (think `led0`). Why is this an alias?

superna9999 replied:
@mbolivar-nordic ok, I just followed what Linux does, busses can be specified as aliases, `chosen` is exclusively use by bootloaders

The DT spec goes in this side

superna9999 replied:
@mbolivar-nordic so according to Zephyr doc, I should a new `zephyr,pcie-controller` Zephyr-specific chosen property ?

superna9999 replied:
OK done

mbolivar-nordic replied:
> @mbolivar-nordic so according to Zephyr doc, I should a new `zephyr,pcie-controller` Zephyr-specific chosen property ?

That's what we generally do when we want to pick a node for use by a subsystem, yeah. Thanks.

superna9999 committed:
doc: devicetree: add PCIe controller chosen property

Add a new chosen property to indentify the main PCIe Controller
node of the system.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 commented:
> * Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
> * Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

PR #38106 is now merged, and this was rebased on main

jhedberg commented on drivers/pcie/host/pcie.c:
```diff
@@ -142,13 +162,32 @@ bool pcie_get_mbar(pcie_bdf_t bdf,
+				    PCIE_CONF_BAR_MEM(phys_addr) ?
+					  PCIE_CONF_BAR_IO_ADDR(phys_addr)
+					: PCIE_CONF_BAR_ADDR(phys_addr),
+				    &mbar->phys_addr))
+		return false;
```
Missing { }

superna9999 replied:
thx

jhedberg commented on drivers/pcie/host/controller.c:
```diff
@@ -0,0 +1,215 @@
+{
+	volatile uint32_t *bdf_cfg_mem = (volatile uint32_t *)((uintptr_t)cfg_addr + (bdf << 4));
+
+	if (!cfg_addr)
+		return;
```
Missing { }

superna9999 replied:
thx

jhedberg commented on drivers/pcie/host/controller.c:
```diff
@@ -0,0 +1,215 @@
+{
+	volatile uint32_t *bdf_cfg_mem = (volatile uint32_t *)((uintptr_t)cfg_addr + (bdf << 4));
+
+	if (!cfg_addr)
+		return 0xffffffff;
```
Missing { }

superna9999 replied:
thx

jhedberg reviewed (changes requested):
A few minor coding style issue regarding missing braces (Zephyr requires them for even single-line branches). What's more strange is that we apparently still don't have an automated check for this in CI?

superna9999 commented:
> * Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
> * Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

@galak is there something blocking ?

cfriedt commented:
🥳
